jib-core/src/main/java/com/google/cloud/tools/jib/registry/RegistryAuthenticator.java

```diff
@@ -263,33 +269,71 @@ URL getAuthenticationUrl(String scope) throws MalformedURLException {
    */
   private Authorization authenticate(String scope) throws RegistryAuthenticationFailedException {
     try {
-      URL authenticationUrl = getAuthenticationUrl(scope);
-
-      try (Connection connection = Connection.getConnectionFactory().apply(authenticationUrl)) {
-        Request.Builder requestBuilder =
-            Request.builder().setHttpTimeout(JibSystemProperties.getHttpTimeout());
-        if (authorization != null) {
-          requestBuilder.setAuthorization(authorization);
-        }
-        Response response = connection.get(requestBuilder.build());
-        String responseString = Blobs.writeToString(response.getBody());
-
-        AuthenticationResponseTemplate responseJson =
-            JsonTemplateMapper.readJson(responseString, AuthenticationResponseTemplate.class);
-        if (responseJson.getToken() == null) {
-          throw new RegistryAuthenticationFailedException(
-              registryEndpointRequestProperties.getServerUrl(),
-              registryEndpointRequestProperties.getImageName(),
-              "Did not get token in authentication response from " + authenticationUrl);
-        }
-        return Authorizations.withBearerToken(responseJson.getToken());
-      }
+      AuthenticationResponseTemplate responseJson =
+          credential != null && credential.isRefreshToken()
+              ? fetchTokenWithOAuth(scope, credential.getPassword())
+              : fetchTokenWithBasicAuth(scope);
 
+      if (responseJson.getToken() == null) {
+        throw new RegistryAuthenticationFailedException(
+            registryEndpointRequestProperties.getServerUrl(),
+            registryEndpointRequestProperties.getImageName(),
+            "Did not get token in authentication response from " + getAuthenticationUrl(scope));
+      }
+      return Authorizations.withBearerToken(responseJson.getToken());
     } catch (IOException ex) {
       throw new RegistryAuthenticationFailedException(
           registryEndpointRequestProperties.getServerUrl(),
           registryEndpointRequestProperties.getImageName(),
           ex);
     }
   }
+
+  private AuthenticationResponseTemplate fetchTokenWithOAuth(String scope, String token)
+      throws IOException, RegistryAuthenticationFailedException {
+    URL authenticationUrl = new URL(realm);
+
+    try (Connection connection = Connection.getConnectionFactory().apply(authenticationUrl)) {
+      BlobHttpContent formBody =
+          new BlobHttpContent(
+              Blobs.from(
+                  "grant_type=refresh_token"
```

Comment:
Probably you can factor this out as a method in the same spirit as `getAuthenticationUrl()` and have a test.

Reply:
Done, please also review my method name